### PR TITLE
gcc 9.1.0

### PR DESCRIPTION
**Pull request by fxcoudert**
Testing the second release candidate for GCC 9.1.0 https://gcc.gnu.org/ml/gcc/2019-04/msg00288.html

### CODE REVIEW

**Comment by fxcoudert** [comment]
@zbeekman do we need a patch for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89864 ? Or was it fixed in 9.1.0?

**Comment by zbeekman** [comment]
I beleive the patch made it into 9.1. GCC won't bootstrap without out it, so if GCC bootstraps and builds, then we're good to go!

File: Formula/gcc.rb [hunk 1]
```diff
@@ -28,16 +27,6 @@ class Gcc < Formula
...
 
-  # Patch for Xcode bug, taken from https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89864#c43
-  # This should be removed in the next release of GCC if fixed by apple; this is an xcode bug,
-  # but this patch is a work around committed to GCC trunk
-  if MacOS::Xcode.version >= "10.2"
-    patch do
-      url "https://raw.githubusercontent.com/Homebrew/formula-patches/master/gcc/8.3.0-xcode-bug-_Atomic-fix.patch"
-      sha256 "33ee92bf678586357ee8ab9d2faddf807e671ad37b97afdd102d5d153d03ca84"
-    end
-  end
```
Comment (zbeekman) [inline Formula/gcc.rb]: LGTM

**Comment by zbeekman** [comment]
Linkage failures needing revision bumps:

- [ ] abyss
- [ ] arpack
- [ ] binwalk
- [ ] cp2k
- [ ] dungeon
- [ ] dynare
- [ ] eccodes
- [ ] fastme
- [ ] fftw
- [ ] grib-api
- [ ] hdf5
- [ ] hdf5@1.8
- [ ] json-fortran
- [ ] kahip
- [ ] lapack
- [ ] libxc
- [ ] lysp
- [ ] minisat
- [ ] mmseqs2
- [ ] mpich
- [ ] netcdf
- [ ] nwchem
- [ ] octave
- [ ] open-mpi
- [ ] openblas
- [ ] packmol
- [ ] petsc
- [ ] petsc-complex
- [ ] pgplot
- [ ] plplot
- [ ] qrupdate
- [ ] r
- [ ] reprepro
- [ ] root
- [ ] scalapack
- [ ] scipy
- [ ] spades

**Comment by zbeekman** [comment]
@fxcoudert this is still a release candidate, correct?

**Comment by fxcoudert** [comment]
GCC 9.1 is released: https://gcc.gnu.org/ml/gcc/2019-05/msg00024.html

File: Formula/gcc.rb (hunk 1 above)
Comment (zbeekman) [reply Formula/gcc.rb]: I spoke too soon. It looks like they're targeting 9.2 for the fix... I'll try to attend to this in ~ 2 hours

**Comment by fxcoudert** [comment]
We'll need support in `brew` itself before this can be merged: https://github.com/Homebrew/brew/pull/6093

**Comment by fxcoudert** [comment]
`brew` support is merged, so we can test this on CI. We will still need to wait until a new `brew` tag is shipped to merge the PR.

**Comment by fxcoudert** [comment]
`openblas` fails to build with

```
==> Downloading from https://codeload.github.com/xianyi/OpenBLAS/tar.gz/v0.3.5
==> make CC=gcc-9 FC=gfortran libs netlib shared
Last 150 lines from /Users/brew/Jenkins/workspace/core/logs/openblas/01.make:
      |           ^~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1082:5: note: in expansion of macro 'KERNEL16x1_SUB'
 1082 |     KERNEL16x1_SUB()
      |     ^~~~~~~~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:55,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx512fintrin.h:3989:1: error: inlining failed in call to always_inline '_mm512_broadcastss_ps': target specific option mismatch
 3989 | _mm512_broadcastss_ps (__m128 __A)
      | ^~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:659:12: note: called from here
  659 |  zmm2   =  _mm512_broadcastss_ps(_mm_load_ss(BO));  \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1082:5: note: in expansion of macro 'KERNEL16x1_SUB'
 1082 |     KERNEL16x1_SUB()
      |     ^~~~~~~~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:55,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx512fintrin.h:6263:1: error: inlining failed in call to always_inline '_mm512_storeu_ps': target specific option mismatch
 6263 | _mm512_storeu_ps (void *__P, __m512 __A)
      | ^~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:669:2: note: called from here
  669 |  _mm512_storeu_ps(CO1      , row0);    \
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1086:4: note: in expansion of macro 'SAVE16x1'
 1086 |    SAVE16x1(alpha)
      |    ^~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:55,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx512fintrin.h:6237:1: error: inlining failed in call to always_inline '_mm512_loadu_ps': target specific option mismatch
 6237 | _mm512_loadu_ps (void const *__P)
      | ^~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:668:11: note: called from here
  668 |  row0  += _mm512_loadu_ps(CO1);     \
      |           ^~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1086:4: note: in expansion of macro 'SAVE16x1'
 1086 |    SAVE16x1(alpha)
      |    ^~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:55,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx512fintrin.h:251:1: error: inlining failed in call to always_inline '_mm512_set1_ps': target specific option mismatch
  251 | _mm512_set1_ps (float __A)
      | ^~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:666:11: note: called from here
  666 |  zmm0   = _mm512_set1_ps(ALPHA);     \
      |           ^~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1086:4: note: in expansion of macro 'SAVE16x1'
 1086 |    SAVE16x1(alpha)
      |    ^~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:51,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avxintrin.h:1227:1: error: inlining failed in call to always_inline '_mm256_setzero_ps': target specific option mismatch
 1227 | _mm256_setzero_ps (void)
      | ^~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:675:9: note: called from here
  675 |  ymm4 = _mm256_setzero_ps();
      |         ^~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1098:4: note: in expansion of macro 'INIT8x1'
 1098 |    INIT8x1()
      |    ^~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:53,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx2intrin.h:934:1: error: inlining failed in call to always_inline '_mm256_broadcastss_ps': target specific option mismatch
  934 | _mm256_broadcastss_ps (__m128 __X)
      | ^~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:679:12: note: called from here
  679 |  ymm2   =  _mm256_broadcastss_ps(_mm_load_ss(BO));  \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1102:5: note: in expansion of macro 'KERNEL8x1_SUB'
 1102 |     KERNEL8x1_SUB()
      |     ^~~~~~~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:51,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avxintrin.h:895:1: error: inlining failed in call to always_inline '_mm256_loadu_ps': target specific option mismatch
  895 | _mm256_loadu_ps (float const *__P)
      | ^~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:678:11: note: called from here
  678 |  ymm0   = _mm256_loadu_ps(AO);     \
      |           ^~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1102:5: note: in expansion of macro 'KERNEL8x1_SUB'
 1102 |     KERNEL8x1_SUB()
      |     ^~~~~~~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:53,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx2intrin.h:934:1: error: inlining failed in call to always_inline '_mm256_broadcastss_ps': target specific option mismatch
  934 | _mm256_broadcastss_ps (__m128 __X)
      | ^~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:679:12: note: called from here
  679 |  ymm2   =  _mm256_broadcastss_ps(_mm_load_ss(BO));  \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1102:5: note: in expansion of macro 'KERNEL8x1_SUB'
 1102 |     KERNEL8x1_SUB()
      |     ^~~~~~~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:51,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avxintrin.h:901:1: error: inlining failed in call to always_inline '_mm256_storeu_ps': target specific option mismatch
  901 | _mm256_storeu_ps (float *__P, __m256 __A)
      | ^~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:689:2: note: called from here
  689 |  _mm256_storeu_ps(CO1      , ymm4);    \
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1106:4: note: in expansion of macro 'SAVE8x1'
 1106 |    SAVE8x1(alpha)
      |    ^~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:51,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avxintrin.h:895:1: error: inlining failed in call to always_inline '_mm256_loadu_ps': target specific option mismatch
  895 | _mm256_loadu_ps (float const *__P)
      | ^~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:688:11: note: called from here
  688 |  ymm4  += _mm256_loadu_ps(CO1);     \
      |           ^~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1106:4: note: in expansion of macro 'SAVE8x1'
 1106 |    SAVE8x1(alpha)
      |    ^~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:51,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avxintrin.h:1310:1: error: inlining failed in call to always_inline '_mm256_set1_ps': target specific option mismatch
 1310 | _mm256_set1_ps (float __A)
      | ^~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:686:11: note: called from here
  686 |  ymm0   = _mm256_set1_ps(ALPHA);     \
      |           ^~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1106:4: note: in expansion of macro 'SAVE8x1'
 1106 |    SAVE8x1(alpha)
      |    ^~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:53,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx2intrin.h:927:1: error: inlining failed in call to always_inline '_mm_broadcastss_ps': target specific option mismatch
  927 | _mm_broadcastss_ps (__m128 __X)
      | ^~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:699:12: note: called from here
  699 |  xmm2   =  _mm_broadcastss_ps(_mm_load_ss(BO));   \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1122:5: note: in expansion of macro 'KERNEL4x1_SUB'
 1122 |     KERNEL4x1_SUB()
      |     ^~~~~~~~~~~~~
In file included from /usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/immintrin.h:53,
                 from ../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:57:
/usr/local/Cellar/gcc/9.1.0/lib/gcc/9/gcc/x86_64-apple-darwin18/9.1.0/include/avx2intrin.h:927:1: error: inlining failed in call to always_inline '_mm_broadcastss_ps': target specific option mismatch
  927 | _mm_broadcastss_ps (__m128 __X)
      | ^~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:699:12: note: called from here
  699 |  xmm2   =  _mm_broadcastss_ps(_mm_load_ss(BO));   \
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../kernel/x86_64/sgemm_kernel_16x4_skylakex.c:1122:5: note: in expansion of macro 'KERNEL4x1_SUB'
 1122 |     KERNEL4x1_SUB()
      |     ^~~~~~~~~~~~~
make[1]: *** [sgemm_kernel_SKYLAKEX.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [libs] Error 1
```

**Comment by fxcoudert** [comment]
The `openblas` build failure should be fixed by updating to 0.3.6: https://github.com/xianyi/OpenBLAS/releases/tag/v0.3.6

but this is blocked by a `visp` build failure: https://github.com/Homebrew/homebrew-core/pull/39438

File: Formula/gcc.rb (hunk 1 above)
Comment (zbeekman) [reply Formula/gcc.rb]: OK, I can confirm that this is in 9.1. Sorry for the confusion. Someone updated the bug report to say it was targeting 9.2 which was the source of my confusion.

**Comment by fxcoudert** [comment]
- `armadillo` and `imake` need revision bumps
- `minisat` is apparently a dead project, it doesn't compile with system compiler (clang) and needs heavy patching (to which upstream is not responding). I'm removing it.

**Comment by zbeekman** [comment]
Was the visp failure ever resolved? When I looked at it pkg-config wasn't helping to get their vendored dependency to build.

**Comment by fxcoudert** [comment]
@zbeekman I resolved it by making visp depend on pcl, which fixes the issue. The bug was only in the codepath taken when pcl is not found.

**Comment by zbeekman** [comment]
Oh, great! I didn't see that pcl was in homebrew. It makes sense that without depending on it it might not build because of how homebrew sanitizes the environment.

**Comment by fxcoudert** [comment]
brew 2.1.2 has shipped with GCC 9 support: https://github.com/Homebrew/brew/releases/tag/2.1.2
so we'll be able to merge this once testing passes

**Comment by fxcoudert** [comment]
The OpenMP issues are due to a conflict between GCC's libgomp and our `libomp` formula shipping `libgomp.{dylib,a}` aliases to its `libomp.{dylib,a}`, despite not being compatible. Will be fixed by updating and now building LLVM's openmp libraries with `LIBOMP_INSTALL_ALIASES=OFF`: https://github.com/Homebrew/homebrew-core/pull/39563/files

**Comment by alexwal** [comment]
Is there a temporary fix in the meantime? This is blocking me :/

**Comment by zbeekman** [comment]
@alexwal what do you mean fix? gcc 9 was just released. How is this blocking you?

The "only" thing we're waiting on now is the CI servers to finish building GCC and all the formula that required revision bumps. The build is in progress and should finish in a few hours if nothing goes wrong.

**Comment by alexwal** [comment]
Ah my mistake. I had a build error installing something with brew and this is where it sent me. Figured the PR was for my issue, but after a closer look, it’s something a little different

**Comment by zbeekman** [comment]
Installing from source I surmise? I’d recommend installing everything you
possibly can as the default bottle builds; it’ll all just work that way.
Only thing I ever build locally is mpich because I want an all GCC based
MPI stack for certain development tasks.

If you hit an error due to OpenMP, the llvm openmp formula will be/was
patched and merged to work around an issue (conflict with gcc) there.

**Comment by fxcoudert** [comment]
GCC 9 has shipped! 🚢

**Pull request closed without merging**